### DISCUSSION TITLE
GetEntityFrameworkEvent() from Web API Service

### DISCUSSION THREAD

**Discussion by stefanos-tsesmetzis** [thepirat000/Audit.NET#546] [Q&A] [answered]
Hi,

I have a .Net Core API in c# that I want to get the EF changes in my service code to store it as json on a DB table.

How can I access the Audit Scope from a server and then call GetEntityFrameworkEvent() to get the event?

Thanks,

**Comment by thepirat000** [comment]
If your only concern is to save the event, you don't need to access the audit scope, but just configure a [data provider](https://github.com/thepirat000/Audit.NET#data-providers) (one of the [data providers included](https://github.com/thepirat000/Audit.NET#data-providers-included) or a custom data provider). Otherwise, please be more specific in what are you trying to accomplish.

Also, if you need to manipulate the audit event before of after the save operation, you can use [custom actions](https://github.com/thepirat000/Audit.NET#custom-actions).

**Reply by stefanos-tsesmetzis** [reply]
I do that already and create nice text log files.

What I'm trying to do is get the same json you are saving in the file and process it to make human readable .
I'm saving that to the DB so present as an update activity for a record.

**Reply by stefanos-tsesmetzis** [reply]
I want to be able to do that from my webapi service because that's where I have access to the previous version of the EF entity and the new one. 

Can I use dependency injection with the AuditScope object?

**Comment by thepirat000** [comment]
No, the audit scope is unaware of any dependency injection mechanism.

Also the EF audit scope cannot be accessed from outside the DbContext before the saving takes place, the only way to access the event before it's saved is with a global custom action.

The library does not provide a standard way to get the saved events, but you can read and deserialize an EF event from a file like this:

```
var ev = JsonConvert.DeserializeObject<AuditEventEntityFramework>(File.ReadAllText("20221127001234_10000.json"));
```

I'm still not sure what are you trying to do. Looks like you want to add some information to the EF audit event after the save operation takes place, but inside the webapi that generated the save operation.

If that's the case, I would suggest having different audit events, one for the webapi and another for EF, but both correlated by the HttpContext's TraceIdentifier. 

You could use [Audit.WebApi](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.WebApi/README.md) extension to add logging to some (or all) your webapi controller methods.

You can see an example of this, by running the following commands:

```
dotnet new -i Audit.WebApi.Template
dotnet new webapiaudit -E
```

It will generate a sample webapi project that includes auditing EF and WebAPI with correlated events.

In summary, you would have to:

- On your startup, add the action filter that generates audits for webapi:

```c#
builder.Services.AddControllers(mvcOptions => 
    // Add the global Audit filter
    mvcOptions.AddAuditFilter(a => a
        .LogAllActions()
        .WithEventType("Event_MVC")) 
);
```

- On your startup, add the http context accessor:

```c#
builder.Services.AddHttpContextAccessor();
```

- On your startup after building the app, add the TraceId custom field to the events

```c#
var httpContextAccessor = app.Services.GetRequiredService<IHttpContextAccessor>();
Audit.Core.Configuration.AddCustomAction(ActionType.OnScopeCreated, scope =>
{
    scope.SetCustomField("TraceId", httpContextAccessor.HttpContext?.TraceIdentifier);
});
```

On your controller, you can access the audit scope for the web api event and add any information you need:

```c#
// POST api/values
[HttpPost]
public async Task<ActionResult<int>> Post([FromBody] string value)
{
    var result = await _provider.InsertAsync(value);

    // Get the current AuditScope instance for the Web API (MVC) event
    var auditScope = this.GetCurrentAuditScope();
            
    // Add an extra field to the Web API event
    auditScope.SetCustomField("ExtraData", new { oldValue = "whatever", newValue = value });

    return Ok(result);
}
```

You will get 2 events per operation, both having the same "TraceId" field

**Reply by stefanos-tsesmetzis** [reply]
Let me try to explain it a bit better:

In my web service, not the controller, I have methods for Create, Update and Delete a record in the DB using EF.
I also have setup AuditDBContext, so when an insert, update or delete happens in EF is recorded to a text file.

BUT, the major issue I have is find a specific history of CRUD for a single record, for example a member's record.

So, would be extremely helpful, if at the Create, Update and Delete methods, I can get the same json Audit.Net writes to the file, so I can save it to an activity DB table, along the MemberID, so I can view the trail of updates to that record.

Is that clear?

**Reply by thepirat000** [reply]
I see...

In that case, can't you have a custom data provider that writes the audit events to both the file and the activity DB table?

```c#
public class CustomDataProvider : Audit.Core.Providers.FileDataProvider
{
	public CustomDataProvider(Action<Audit.Core.ConfigurationApi.IFileLogProviderConfigurator> config) : base(config)
	{}

	public override object InsertEvent(AuditEvent auditEvent)
	{
		var id = base.InsertEvent(auditEvent);

		// Get the EF event and insert to the activity DB
		var efEvent = auditEvent.GetEntityFrameworkEvent();
		// ...

		return id;
	}

	public override async Task<object> InsertEventAsync(AuditEvent auditEvent)
	{
		var id = await base.InsertEventAsync(auditEvent);

		// Get the EF event and insert to the activity DB
		var efEvent = auditEvent.GetEntityFrameworkEvent();
		// ...

		return id;
	}
}
```

**Reply by stefanos-tsesmetzis** [reply]
The secret though is to insert the event activity to the right table.
For example: for a Members table I have a MembersActivity one that needs a MEMBERID field to associate it with its master record.

That's why I want to do it at the service level, because I have all the PKs there.

**Comment by stefanos-tsesmetzis** [comment]
Sorry for being a pain.

_dbContext.AddAuditCustomField("ExtraField", new { InsertedId =
added.Entity.Id, ... });

will add a new field in the event that is written in the text file.

I want to get the actual json that contains the entity changes (fields) and
save it to the proper DB table (MembersActivity, OrdersActivity eta).


Stefanos Tsesmetzis Founder
o: (516) 216-9474  c: (917) 214-0938
logismico.com <http://www.logismico.com>

**Reply by thepirat000** [reply]
You're not a pain, don't worry :)
Yeah, as soon as I posted my previous answer I realized that was not what you were looking for.

Can you share the service code where you update the entity and want to get the audit event?

**Comment by stefanos-tsesmetzis** [comment]
There you go:

            var member = await _context.Members
                                            .FirstOrDefaultAsync(x =>
x.MemberPublicId == request.MemberPublicId, ct);

            if (member == null) throw new
RestException(HttpStatusCode.NotAcceptable, "Member is not found");

            member.IsDeleted = request.isDeleted;
            member.FirstName = request.FirstName;
            member.LastName = request.LastName;
            member.Type = _context.Types.Where(x => x.TypeName ==
request.Type).FirstOrDefault();
            member.Notes = request.Notes;
            member.Status = request.Status;

            using var transaction = _context.Database.BeginTransaction();

            try
            {

//------------------------------------------------------------
                // Record Activity
                _context.MemberActivities.Add(
                    new MemberActivity()
                    {
                        ActivityType = AppConstant.ActivityUpdate,
                        Dated = DateTime.UtcNow,
                        MemberId = member.MemberId,
                        Notes = *{ I WANT TO ADD THE CHAMGES FROM THE
ORIGINAL MEMBER RECORD HERE},   <=================================*
                        MemberActivityPublicId = Guid.NewGuid()
                    }
                    ); ;

                //--------------------------------
                await _context.SaveChangesAsync(ct);
                transaction.Commit();
                //--------------------------------

                response.PublicId = member.MemberPublicId;
                return response;
            }
            catch (Exception e)
            {
                transaction.Rollback();

                _logger.LogError(e.Message);
                throw new RestException(HttpStatusCode.InternalServerError,
e.Message);
            }

Stefanos Tsesmetzis Founder
o: (516) 216-9474  c: (917) 214-0938
logismico.com <http://www.logismico.com>

**Reply by thepirat000** [reply]
does your DbContext inherits from AuditDbContext or do you use the AuditSaveChanges interceptor?

**Comment by stefanos-tsesmetzis** [comment]
DbContext inherits from AuditDbContext.

Stefanos Tsesmetzis
***@***.***

**Comment by thepirat000** [comment]
I have an idea to provide a way to obtain the EntityFrameworkEvent when calling SaveChanges, but it implies using the EF audit **without inheritance** like in this example:
https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#2-without-inheritance-overriding-savechanges

So the `_helper` will provide another method to Save Changes, but this new method will return the EF event instead of only the int result.

So your DbContext can provide a new "SaveChangesAudited()" that saves the changes and returns the generated audit event, by calling the new method in the helper.

So your service can call this new method on your DbContext instead:

```c#
try
            {

//------------------------------------------------------------

                //Calling the new savechanges
                var efEvent = await _context.SaveChangesAuditedAsync(ct);  // <---- calling the new save changes

                // Record Activity (AFTER saving changes)
                _context.MemberActivities.Add(
                    new MemberActivity()
                    {
                        Notes = efEvent.ToJson(), // <----- here
                        ...
                    }) ;

                // you will probably need to call savechanges here to save the activity, but it will execute in the same transaction
               await _context.SaveChangesAuditedAsync(ct);

                transaction.Commit();
                //--------------------------------

                response.PublicId = member.MemberPublicId;
                return response;
            }
            catch (Exception e)
            {
                transaction.Rollback();

                _logger.LogError(e.Message);
                throw new RestException(HttpStatusCode.InternalServerError,
e.Message);
            }
```

**Comment by stefanos-tsesmetzis** [comment]
Federico,

This example is not providing a SaveChangesAuditedAsync() method.

[image: image.png]


Stefanos Tsesmetzis Founder
o: (516) 216-9474  c: (917) 214-0938
logismico.com <http://www.logismico.com>

**Comment by thepirat000** [comment]
I didn't include, since it's just an idea so far and the helper doesn't have the method that returns the audit event (i.e. `SaveChangesGetAudit`), but it will look something like this:

```c#
public class MyContext : DbContext
{
    private readonly DbContextHelper _helper = new DbContextHelper();
    private readonly IAuditDbContext _auditContext;
        
    public MyContext(DbContextOptions<MyContext> options) : base(options)
    {
        _auditContext = new DefaultAuditContext(this);
        _helper.SetConfig(_auditContext);
    }

    // ...  

    public EntityFrameworkEvent SaveChangesAudited(bool acceptAllChangesOnSuccess = true)
    {
        return _helper.SaveChangesGetAudit(_auditContext, () => base.SaveChanges(acceptAllChangesOnSuccess));
    }

    public Task<EntityFrameworkEvent> SaveChangesAuditedAsync(bool acceptAllChangesOnSuccess = true, CancellationToken cancellationToken = default)
    {
        return _helper.SaveChangesGetAuditAsync(_auditContext, () => base.SaveChangesAsync(acceptAllChangesOnSuccess));
    }
        
    public override int SaveChanges(bool acceptAllChangesOnSuccess)
    {
        throw new NotImplementedException("Must call SaveChangesAudited method");
    }

    public override Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException("Must call SaveChangesAuditedAsync method");
    }
}
```

Anyway I will take a closer look, maybe it can be done also when inheriting from AuditDbContext.

**Comment by stefanos-tsesmetzis** [comment]
This will be a great feature for you guys too.

Many times in data-sensitive applications, developers have to present the
user with the updates on a record in a human-readable format.

Like  this library
https://github.com/GregFinzer/Compare-Net-Objects
And a feature they have called UserFriendlyReport

It is not working well with nested children.

Stefanos Tsesmetzis Founder
o: (516) 216-9474  c: (917) 214-0938
logismico.com <http://www.logismico.com>

**Comment by thepirat000** [comment]
Will follow up in #547

**Comment by thepirat000** [comment, marked as answer]
This was implemented in version 20.0.4, please upgrade your references and re-test.

If your DbContext inherits from AuditDbContext, you can use `SaveChangesGetAudit()` or `SaveChangesGetAuditAsync()` to apply the changes and obtain the audit event generated:

```c#
using var ctx = new MyContext();
EntityFrameworkEvent efEvent = ctx.SaveChangesGetAudit();
```

**Comment by stefanos-tsesmetzis** [comment]
It works like charm.

Thank you Federico.

Stefanos Tsesmetzis Founder
o: (516) 216-9474  c: (917) 214-0938
logismico.com <http://www.logismico.com>
